# Finding our best-performing salespeople and products

## Introduction

**Business Context.** You work for AdventureWorks, a company that sells outdoor sporting equipment. The company has many different locations and has been recording the sales of different locations on various products. You, their new data scientist, have been tasked with the question: **"What are our best products and salespeople and how can use this information to improve our overall performance?"**

You have been given access to the relevant data files with documentation from the IT department. Your job is to extract meaningful insights from these data files to help increase sales. First, you will look at the best products and try to see how different products behave in different categories. Second, you will analyze the best salespeople to see if the commission percentage motivates them to sell more.

**Business Problem.** Your task is to **construct a database from the provided CSV files and then write queries in SQL to carry out the requested analysis**.

**Analytical Context.** You are given the data (stored in the ```data/csvs``` folder) as a set of separate CSV files, each one representing a table. You will build a new PostgreSQL database from these files using AWS RDS.

The company has been pretty vague about how they expect you to extract insights, but you have come up with the following plan of attack:

1. Create the database and ensure you can run basic queries against it
2. Look at how product ratings and total sales are related
3. See how products sell in different subcategories (bikes, helmets, socks, etc.)
4. Calculate which salespeople have performed the best in the past year
5. Seeing if total sales are correlated with their commission percentage

Of course, this is only your initial plan. As you explore the database, your strategy will change.

## Setting up AWS

In this case, we'll assume that the company has given you an entry-level laptop, which is not capable of running a PostgreSQL server locally. Therefore, you should set up a cloud database, connect to it from `psql`, and run the analysis via the `psql` or directly from the notebook.

### Question:

Repeat the steps in Case 12.3 to create a new RDS instance with a PostgreSQL database.

## Overview of the data

The data for the case is contained in the ```./data/csvs``` directory; specifically, it is the ```AdventureWorks``` sample data provided by Microsoft. We will be focusing on the Sales and Production categories. Complete documentation for the original data (of which you have only a subset) can be found [here](https://dataedo.com/download/AdventureWorks.pdf). 

**Product Tables:**
* **Product**: one row per product that the company sells
* **ProductReview**: one row per rating and review left by customers
* **ProductModelProductDescriptionCulture**: a link between products and their longer descriptions also indicating a "culture" - which language and region the product is for
* **ProductDescription**: a longer description of each product, for a specific region
* **ProductCategory**: the broad categories that products fit into
* **ProductSubCategory**: the narrower subcategories that products fit into

**Sales Tables:**
* **SalesPerson**: one row per salesperson, including information on their commission and performance
* **SalesOrderHeader**: one row per sale summarizing the sale
* **SalesOrderDetail**: many rows per sale, detailing each product that forms part of the sale
* **SalesTerritory**: the different territories where products are sold, including performance

**Region Tables:**
* **CountryRegionCurrency**: the currency used by each region
* **CurrencyRate**: the average and closing exchange rates for each currency compared to the USD

## Using `ipython-sql` and `pgspecial`

Jupyter notebook is usually used to run Python code, but with an add-on it can run SQL directly against a database too. The extensions `ipython-sql` and `pgspecial` will let you do this.

Load the sql add-on and connect to the database as follows. You'll need to change the username (`postgres`), password (`mysecretpassword`), host (`localhost`), and database name (`postgres`) to what you used when setting up your RDS instance:

In [1]:
%load_ext sql
%sql postgresql://postgres:cYlQurab5M8cAtDvwRNZ@ec3-database.cfnfrwra7yoo.us-east-2.rds.amazonaws.com/adventuretime

You should now be able to run SQL directly from any Jupyter notebook cell by starting the cell with a line that states `%%sql`. For example (once you have a database with some tables, which we'll only create later):

```sql
%%sql

SELECT * FROM product LIMIT 10;
```

**Note:** Unlike `pandas` which automatically truncates output for large DataFrames, the SQL plug-in gives you exactly what you ask for. If you do a `SELECT * FROM` a table with a million rows and no `LIMIT` clause, it'll output all million rows and probably freeze your notebook. It's good practice to always use a `LIMIT` clause even when it's not needed to avoid any mishaps.

## Creating the database and adding the tables

Now, let's create a database called `adventuretime`. (If you do this through the notebook, you'll have to add the line `end;` before your `create database` command as the add-on runs everything in transactions).

You'll need to add a table for each of the CSV files. Spend some time looking at the different CSV files and getting used to how they reference each other and what headers they create. Then, you'll need to write an appropriate `CREATE TABLE` command with appropriate types. You can figure out the types by inspecting the CSV files and/or referencing the documentation.

### Exercise 1:

Write all of the commands that you need to

* Create the database
* Create the tables
* Import the data from the CSVs

**Hint:** As an example, to add data for the `salesperson` table, you would use the following commands:

1. Create table (can be run from Jupyter Notebook or the `psql` command line interface):
```sql
CREATE TABLE salesperson (
    businessentityid INTEGER,
    territoryid INTEGER,
    salesquota INTEGER,
    bonus INTEGER,
    commissionpct FLOAT,
    salesytd FLOAT,
    saleslastyear FLOAT,
    rowguid TEXT,
    modifieddate DATE
    );
```

2. Copy data (has to be run from the `psql` shell):

```sql
\copy salesperson FROM 'data/csvs/salesperson.csv' with (format CSV, header true, delimiter ',');
```

**Answer.**

**<ins>Each of these steps was done directly from the `psql` environment on the terminal. This was done like this to avoid increasing the tables content each time the code is executed here on the notebook.</ins>**<br><br>

1. Creation of the database:
**Establishing the initial connection to the AWS RDS:**
~~~bash
psql -h ec3-database.cfnfrwra7yoo.us-east-2.rds.amazonaws.com -U postgres
password: cYlQurab5M8cAtDvwRNZ
~~~
**Creating the database, user and privileges:**
~~~SQL
create database adventuretime;
create user ec3_user with login encrypted password 'mypassword';
grant all privileges on database adventuretime to ec3_user;
\q
~~~
**Establishing the user connection to the AWS RDS:**
~~~bash
psql -h ec3-database.cfnfrwra7yoo.us-east-2.rds.amazonaws.com -U ec3_user -d adventuretime
password: mypassword
~~~

2. Creation of the tables:
**Product Tables** <br><br>
*Table: Product*
~~~SQL
create table product(
    productid INT primary key,
    name TEXT,
    productnumber TEXT,
    makeflag BOOLEAN,
    finishedgoodsflag BOOLEAN,
    color TEXT,
    safetystocklevel SMALLINT,
    reorderpoint SMALLINT,
    standardcost FLOAT,
    listprice FLOAT,
    size TEXT,
    sizeunitmeasurecode TEXT,
    weightunitmeasurecode TEXT,
    weight FLOAT,
    daystomanufacture INT,
    productline TEXT,
    class TEXT,
    style TEXT,
    productsubcategoryid INT,
    productmodelid INT,
    sellstartdate DATE,
    sellenddate DATE,
    discontinueddate DATE,
    rowguid TEXT,
    modifieddate DATE
);
~~~
*Table: Product Category*
~~~SQL
create table productcategory(
    productcategoryid INT primary key,
    name TEXT,
    rowguid TEXT,
    modifieddate DATE
);
~~~
*Table: Product Description*
~~~SQL
create table productdescription(
    productdescriptionid INT primary key,
    description TEXT,
    rowguid TEXT,
    modifieddate DATE
);
~~~
*Table: Product Model-Description-Culture*
~~~SQL
create table productmodelproductdescriptionculture(
    productmodelid INT,
    productdescriptionid INT,
    cultureid TEXT,
    modifieddate DATE
);
~~~
*Table: Product Review*
~~~SQL
create table productreview(
    productreviewid INT primary key,
    productid INT,
    reviewername TEXT,
    reviewdate DATE,
    emailaddress TEXT,
    rating INT,
    comments TEXT,
    modifieddate DATE
);
~~~
*Table: Product Sub-Category*
~~~SQL
create table productsubcategory(
    productsubcategoryid INT primary key,
    productcategoryid INT,
    name TEXT,
    rowguid TEXT,
    modifieddate DATE
);
~~~
**Sales Tables**<br><br> 
*Table: Country-Region-Currency*
~~~SQL
create table countryregioncurrency(
    countryregioncode TEXT,
    currencycode TEXT,
    modifieddate DATE
);
~~~
*Table: Sales Currency Rate*
~~~SQL
create table currencyrate(
    currencyrateid INT primary key,
    currencyratedate DATE,
    fromcurrencycode TEXT,
    tocurrencycode TEXT,
    averagerate FLOAT,
    endofdayrate FLOAT,
    modifieddate DATE
);
~~~
*Table: Sales Order Header*
~~~SQL
create table salesorderheader(
    salesorderid INT primary key,
    revisionnumber SMALLINT,
    orderdate DATE,
    duedate DATE,
    shipdate DATE,
    status SMALLINT,
    onlineorderflag BOOLEAN,
    purchaseordernumber TEXT,
    accountnumber TEXT,
    customerid INT,
    salespersonid INT,
    territoryid INT,
    billtoaddressid INT,
    shiptoaddressid INT,
    shipmethodid INT,
    creditcardid INT,
    creditcardapprovalcode TEXT,
    currencyrateid INT,
    subtotal FLOAT,
    taxamt FLOAT,
    freight FLOAT,
    totaldue FLOAT,
    comment TEXT,
    rowguid TEXT,
    modifieddate DATE
);
~~~
*Table: Sales Order Detail*
~~~SQL
create table salesorderdetail(
    salesorderid INT,
    salesorderdetailid INT primary key,
    carriertrackingnumber TEXT,
    orderqty SMALLINT,
    productid INT,
    specialofferid INT,
    unitprice FLOAT,
    unitpricediscount FLOAT,
    rowguid TEXT,
    modifieddate DATE
);
~~~
*Table: Salesperson*
~~~SQL
create table salesperson(
    businessentityid INT primary key,
    territoryid INT,
    salesquota FLOAT,
    bonus FLOAT,
    commissionpct FLOAT,
    salesytd FLOAT,
    saleslastyear FLOAT,
    rowguid TEXT,
    modifieddate DATE
);
~~~
*Table: Sales Territory*
~~~SQL
create table salesterritory(
    territoryid INT primary key,
    name TEXT,
    countryregioncode TEXT,
    "group" TEXT,
    salesytd FLOAT,
    saleslastyear FLOAT,
    costytd FLOAT,
    costlastyear FLOAT,
    rowguid TEXT,
    modifieddate DATE
);
~~~

3. Importing the csv data to the previously created tables:

**Product Tables**<br><br>
*Table: Product*
~~~bash
psql -h ec3-database.cfnfrwra7yoo.us-east-2.rds.amazonaws.com -U ec3_user -d adventuretime -c "\copy product FROM 'data/csvs/product.csv' with (format CSV, header true, delimiter ',');"
~~~
*Table: Product Category*
~~~bash
psql -h ec3-database.cfnfrwra7yoo.us-east-2.rds.amazonaws.com -U ec3_user -d adventuretime -c "\copy productcategory FROM 'data/csvs/productcategory.csv' with (format CSV, header true, delimiter ',');"
~~~
*Table: Product Description*
~~~bash
psql -h ec3-database.cfnfrwra7yoo.us-east-2.rds.amazonaws.com -U ec3_user -d adventuretime -c "\copy productdescription FROM 'data/csvs/productdescription.csv' with (format CSV, header true, delimiter ',');"
~~~
*Table: Product Model-Description-Culture*
~~~bash
psql -h ec3-database.cfnfrwra7yoo.us-east-2.rds.amazonaws.com -U ec3_user -d adventuretime -c "\copy productmodelproductdescriptionculture FROM 'data/csvs/productmodelproductdescriptionculture.csv' with (format CSV, header true, delimiter ',');"
~~~
*Table: Product Review*
~~~bash
psql -h ec3-database.cfnfrwra7yoo.us-east-2.rds.amazonaws.com -U ec3_user -d adventuretime -c "\copy productreview FROM 'data/csvs/productreview.csv' with (format CSV, header true, delimiter ',');"
~~~
*Table: Product Sub-Category*
~~~bash
psql -h ec3-database.cfnfrwra7yoo.us-east-2.rds.amazonaws.com -U ec3_user -d adventuretime -c "\copy productsubcategory FROM 'data/csvs/productsubcategory.csv' with (format CSV, header true, delimiter ',');"
~~~
**Sales Tables**<br><br>
*Table: Country-Region-Currency*
~~~bash
psql -h ec3-database.cfnfrwra7yoo.us-east-2.rds.amazonaws.com -U ec3_user -d adventuretime -c "\copy countryregioncurrency FROM 'data/csvs/countryregioncurrency.csv' with (format CSV, header true, delimiter ',');"
~~~
*Table: Sales Currency Rate*
~~~bash
psql -h ec3-database.cfnfrwra7yoo.us-east-2.rds.amazonaws.com -U ec3_user -d adventuretime -c "\copy currencyrate FROM 'data/csvs/currencyrate.csv' with (format CSV, header true, delimiter ',');"
~~~
*Table: Sales Order Header*
~~~bash
psql -h ec3-database.cfnfrwra7yoo.us-east-2.rds.amazonaws.com -U ec3_user -d adventuretime -c "\copy salesorderheader FROM 'data/csvs/salesorderheader.csv' with (format CSV, header true, delimiter ',');"
~~~
*Table: Sales Order Detail*
~~~bash
psql -h ec3-database.cfnfrwra7yoo.us-east-2.rds.amazonaws.com -U ec3_user -d adventuretime -c "\copy salesorderdetail FROM 'data/csvs/salesorderdetail.csv' with (format CSV, header true, delimiter ',');"
~~~
*Table: Salesperson*
~~~bash
psql -h ec3-database.cfnfrwra7yoo.us-east-2.rds.amazonaws.com -U ec3_user -d adventuretime -c "\copy salesperson FROM 'data/csvs/salesperson.csv' with (format CSV, header true, delimiter ',');"
~~~
*Table: Sales Territory*
~~~bash
psql -h ec3-database.cfnfrwra7yoo.us-east-2.rds.amazonaws.com -U ec3_user -d adventuretime -c "\copy salesterritory FROM 'data/csvs/salesterritory.csv' with (format CSV, header true, delimiter ',');"
~~~

-------

## Finding our most popular products

As discussed, the company would like to know which of their products is the most popular among customers. You figure that the average rating given in reviews is correlated with the number of sales of a particular product (that products with higher reviews have more sales).

### Exercise 2:

Using the ```product``` and ```productreview``` tables, ```JOIN``` them and rank the products according to their average review rating. What are the names and IDs of the top 5 products?

**Answer.**

In [2]:
%%sql
SELECT product.productid, product.productmodelid, product.name, AVG(productreview.rating) AS avg_rating FROM product
JOIN productreview ON product.productid = productreview.productid
GROUP BY product.productid
ORDER BY avg_rating DESC
LIMIT 5;

 * postgresql://postgres:***@ec3-database.cfnfrwra7yoo.us-east-2.rds.amazonaws.com/adventuretime
3 rows affected.


[(709, 18, 'Mountain Bike Socks, M', Decimal('5.0000000000000000')),
 (798, 29, 'Road-550-W Yellow, 40', Decimal('5.0000000000000000')),
 (937, 64, 'HL Mountain Pedal', Decimal('3.0000000000000000'))]

-------

### Exercise 3:

Much to your disappointment, there are only three products with ratings and only four reviews in total! This is nowhere near enough to perform an analysis of the correlation between reviews and total sales.

Nevertheless, your manager wants the **English description** of these products for an upcoming sale. Use the documentation provided above if you need help navigating the structure to extract this!

**Hint:** You'll notice that the value for `cultureid` in the `productmodelproductdescriptionculture` table often has extra trailing spaces which makes it difficult to reliably get descriptions of a specific language. You should first modify this table before writing the `SELECT` statement to get the descriptions that your manager wants. To do this, you can use an `UPDATE` statement with Postgres's [`TRIM`](https://w3resource.com/PostgreSQL/trim-function.php) function.

**Answer.**

In [3]:
%%sql
SELECT productid, name, description FROM(
    SELECT productmodelproductdescriptionculture.productmodelid, TRIM(BOTH FROM cultureid) AS cultureid_mod, productdescription.description FROM productmodelproductdescriptionculture
    JOIN productdescription ON productdescription.productdescriptionid = productmodelproductdescriptionculture.productdescriptionid
)a
JOIN product ON product.productmodelid = a.productmodelid
WHERE (cultureid_mod = 'en' AND (productid = 709 OR productid = 798 OR productid = 937))
LIMIT 10;

 * postgresql://postgres:***@ec3-database.cfnfrwra7yoo.us-east-2.rds.amazonaws.com/adventuretime
3 rows affected.


[(709, 'Mountain Bike Socks, M', 'Combination of natural and synthetic fibers stays dry and provides just the right cushioning.'),
 (798, 'Road-550-W Yellow, 40', 'Same technology as all of our Road series bikes, but the frame is sized for a woman.  Perfect all-around bike for road or racing.'),
 (937, 'HL Mountain Pedal', 'Stainless steel; designed to shed mud easily.')]

-------

### Exercise 4:

Since we cannot infer the most popular products from the reviews, we will go with an alternative strategy.

Get the model ID, name, description, and total number of sales for each product and display the top-10 selling products. You can infer how often products have been sold by looking at the `salesorderdetail` table (each row might indicate more than one sale, so take note of `OrderQty`).

**Answer.**

In [4]:
%%sql
SELECT b.productmodelid, b.name, description, b.sum_orderqty FROM (
    SELECT a.productmodelid, a.name, a.sum_orderqty,productmodelproductdescriptionculture.productdescriptionid, TRIM(BOTH FROM cultureid) AS cultureid_mod FROM (
        SELECT product.productmodelid, product.name, SUM(salesorderdetail.orderqty) AS sum_orderqty FROM salesorderdetail
        JOIN product ON product.productid = salesorderdetail.productid
        GROUP BY product.productid
        ORDER BY sum_orderqty DESC
    )a
    LEFT JOIN productmodelproductdescriptionculture ON productmodelproductdescriptionculture.productmodelid = a.productmodelid
)b
JOIN productdescription ON productdescription.productdescriptionid = b.productdescriptionid 
WHERE cultureid_mod = 'en'
LIMIT 10;

 * postgresql://postgres:***@ec3-database.cfnfrwra7yoo.us-east-2.rds.amazonaws.com/adventuretime
10 rows affected.


[(2, 'AWC Logo Cap', 'Traditional style with a flip-up brim; one-size fits all.', 8311),
 (111, 'Water Bottle - 30 oz.', 'AWC logo water bottle - holds 30 oz; leak-proof.', 6815),
 (33, 'Sport-100 Helmet, Blue', 'Universal fit, well-vented, lightweight , snap-on visor.', 6743),
 (11, 'Long-Sleeve Logo Jersey, L', 'Unisex long-sleeve AWC logo microfiber cycling jersey', 6592),
 (33, 'Sport-100 Helmet, Black', 'Universal fit, well-vented, lightweight , snap-on visor.', 6532),
 (33, 'Sport-100 Helmet, Red', 'Universal fit, well-vented, lightweight , snap-on visor.', 6266),
 (1, 'Classic Vest, S', 'Light-weight, wind-resistant, packs to fit into a pocket.', 4247),
 (114, 'Patch Kit/8 Patches', 'Includes 8 different size patches, glue and sandpaper.', 3865),
 (32, 'Short-Sleeve Classic Jersey, XL', 'Short sleeve classic breathable jersey with superior moisture control, front zipper, and 3 back pockets.', 3864),
 (11, 'Long-Sleeve Logo Jersey, M', 'Unisex long-sleeve AWC logo microfiber cycling jersey', 3636)]

-------

### Exercise 5:

Let's look at the correlation between quantity sold and price for each item in each subcategory. Some subcategories don't have enough sales to make the correlation meaningful, so only look at the top 10 subcategories by total quantity of sales.

Once you've looked at the data, make a hypothesis about what causes any positive or negative correlations between price and quantity, and explain this in 2-3 sentences.

**Hint:** You'll need to calculate the total quantities from `salesorderdetail` again and group the products by subcategory. It'll probably be easier if you use at least two [CTEs](https://www.postgresql.org/docs/9.1/queries-with.html). You can calculate the correlation in PostgreSQL by using the built-in [```corr()```](https://www.postgresql.org/docs/9.4/functions-aggregate.html) function.

**Answer.**

In [5]:
%%sql
WITH subcategories AS (
    SELECT salesorderdetail.productid, product.productsubcategoryid, orderqty FROM salesorderdetail
    JOIN product ON salesorderdetail.productid = product.productid
), top_10_subcategories AS (
    SELECT productsubcategoryid, SUM(orderqty) AS tot_sales FROM subcategories
    GROUP BY productsubcategoryid
    ORDER BY tot_sales DESC
    LIMIT 10
), corr_table AS (
    SELECT product.productsubcategoryid, product.name, salesorderdetail.productid, salesorderdetail.orderqty, salesorderdetail.unitprice FROM salesorderdetail
    JOIN product ON salesorderdetail.productid = product.productid
    WHERE product.productsubcategoryid IN (SELECT productsubcategoryid FROM top_10_subcategories)
)

SELECT productsubcategoryid, corr(orderqty,unitprice) FROM corr_table
GROUP BY productsubcategoryid
LIMIT 15;

 * postgresql://postgres:***@ec3-database.cfnfrwra7yoo.us-east-2.rds.amazonaws.com/adventuretime
10 rows affected.


[(12, -0.13597349540568923),
 (31, -0.6844422610056845),
 (21, -0.5896023220461615),
 (14, -0.050832548006872),
 (3, -0.23334387310991903),
 (37, -0.07767311985839663),
 (28, -0.2896268378481839),
 (20, -0.30376872837951174),
 (1, -0.22939105049987216),
 (2, -0.28502426085920524)]

Looking at the results we can see that for all subcategories there is a negative correlation between price and quantities. This is what we would've expect to happen, since the demand for most (but no all) goods have a downward slope. In other words, when its price goes up, its demand, and hence the quantities sold, goes down. This is what is commonly known in economics as the [consumer demand theory](https://www.investopedia.com/terms/d/demand_theory.asp).

-------

## Finding our top salespeople

As mentioned earlier, we want to find our best salespeople and see whether or not we can incentivize them in an appropriate manner. Namely, we want to determine if the commission percentage we give them motivates them to make more and bigger sales.

### Exercise 6:

Find the top five performing salespeople by using the `salesytd` (Sales, year-to-date) column. (We only need to know the `businessentityid` for each salesperson as this uniquely identifies each.) Why might you be skeptical of these numbers right now?

**Answer.**

In [6]:
%%sql
SELECT businessentityid, salesytd FROM salesperson
ORDER BY salesytd DESC
LIMIT 5;

 * postgresql://postgres:***@ec3-database.cfnfrwra7yoo.us-east-2.rds.amazonaws.com/adventuretime
5 rows affected.


[(276, 4251368.5497),
 (289, 4116871.2277),
 (275, 3763178.1787),
 (277, 3189418.3662),
 (290, 3121616.3202)]

One might think that the comparison we are doing is not well specified. For instance, we are not currently controlling for the number of years each salesperson has performed in the company, so this could lead to a potentially bad classification of the best performing salespeople.

-------

### Exercise 7:

Using ```salesorderheader```, find the top 5 salespeople who made the most sales **in the most recent year** (2014). (There is a column called `subtotal` - use that.) Sales that do not have an associated salesperson should be excluded from your calculations and final output. All orders that were made within the 2014 calendar year should be included.

**Hint:** You can use the syntax `'1970-01-01'::date` to generate an arbitrary date in PostgreSQL and compare this to specific dates in the tables.

**Answer.**

In [7]:
%%sql
WITH sales_2014 AS(
    SELECT TO_CHAR(orderdate,'YYYY') AS order_year, salespersonid, subtotal FROM salesorderheader
    WHERE (TO_CHAR(orderdate,'YYYY') = '2014' AND salespersonid IS NOT NULL)
)

SELECT salespersonid, SUM(subtotal) AS tot_sales FROM sales_2014
GROUP BY salespersonid
ORDER BY tot_sales DESC
LIMIT 5;

 * postgresql://postgres:***@ec3-database.cfnfrwra7yoo.us-east-2.rds.amazonaws.com/adventuretime
5 rows affected.


[(289, 1382996.5838999997),
 (276, 1271088.5216000006),
 (275, 1057247.3785999997),
 (282, 1044810.8277000004),
 (277, 1040093.4071000002)]

-------

### Exercise 8:

Looking at the documentation, you will see that `subtotal` in the ```salesorderheader``` table is calculated from other tables in the database. To validate this figure (instead of trusting it blindly), let's calculate `subtotal` manually. Using the ```salesorderdetail``` and ```salesorderheader``` tables, calculate the sales for each salesperson for **this past year** (2014) and display results for the top 5 salespeople.

**Hint:** You will have to ```JOIN``` ```salesorderdetail``` on ```salesorderheader``` to get the salesperson, calculate line totals for each sale using appropriate discounts, then sum all the line totals to get the total sale. You will want to use ```WITH``` clauses again to keep things sane.

**Answer.**

In [8]:
%%sql
WITH new_salesorderdetail AS(
    SELECT salesorderid, COALESCE((unitprice * (1-unitpricediscount) * orderqty),0) AS linetotal FROM salesorderdetail
), sales_2014 AS (
    SELECT TO_CHAR(orderdate,'YYYY') AS order_year, salesorderheader.salesorderid, salesorderheader.salespersonid, salesorderheader.subtotal, new_salesorderdetail.linetotal FROM salesorderheader
    JOIN new_salesorderdetail ON new_salesorderdetail.salesorderid = salesorderheader.salesorderid
    WHERE (TO_CHAR(orderdate,'YYYY') = '2014' AND salespersonid IS NOT NULL) 
)

SELECT salespersonid, SUM(linetotal) AS tot_sales FROM sales_2014
GROUP BY salespersonid
ORDER BY tot_sales DESC
LIMIT 5;

 * postgresql://postgres:***@ec3-database.cfnfrwra7yoo.us-east-2.rds.amazonaws.com/adventuretime
5 rows affected.


[(289, 1382996.5839099996),
 (276, 1271088.5214609967),
 (275, 1057247.378571998),
 (282, 1044810.8276869985),
 (277, 1040093.4069009989)]

As we would expect, the results from the manual calculation are the same as those that were obtained directly using the `subtotal` variable. This is consistent with what is contained in the documentation of the database.

-------

### Exercise 9:

Using ```corr()```, see if there is a positive relationship between total sales and commission percentage.

**Answer.**

In [9]:
%%sql
WITH new_salesorderdetail AS(
    SELECT salesorderid, COALESCE((unitprice * (1-unitpricediscount) * orderqty),0) AS linetotal FROM salesorderdetail
), sales_2014 AS (
    SELECT TO_CHAR(orderdate,'YYYY') AS order_year, salesorderheader.salesorderid, salesorderheader.salespersonid, salesorderheader.subtotal, new_salesorderdetail.linetotal FROM salesorderheader
    JOIN new_salesorderdetail ON new_salesorderdetail.salesorderid = salesorderheader.salesorderid
    WHERE (TO_CHAR(orderdate,'YYYY') = '2014' AND salespersonid IS NOT NULL) 
)

SELECT salespersonid, corr(linetotal,commissionpct) FROM sales_2014
JOIN salesperson ON salesperson.businessentityid = sales_2014.salespersonid
GROUP BY salespersonid
ORDER BY corr DESC
LIMIT 20;

 * postgresql://postgres:***@ec3-database.cfnfrwra7yoo.us-east-2.rds.amazonaws.com/adventuretime
17 rows affected.


[(274, None),
 (287, None),
 (285, None),
 (288, 0.06557470024520988),
 (279, 0.06002820655578976),
 (284, 0.05160509576241608),
 (290, 0.04840518442019965),
 (286, 0.010331163888262076),
 (275, -0.005361103034878941),
 (280, -0.01500803527864619),
 (289, -0.015312929264963492),
 (283, -0.016579385522833837),
 (278, -0.047484554610116796),
 (277, -0.04780165517387576),
 (282, -0.05648994786032343),
 (276, -0.0705920955740888),
 (281, -0.09627261295133765)]

Indeed there are some cases where we could see a positive correlation between sales and the commission percentage for each salesperson. However, the resulting values are very small, being the maximum 6%. On the other hand, there are several negative correlations, and even some salespersons that do not exhibit any correlation at all. 

-------

### Exercise 10:

Remember how we mentioned that products were sold in many regions? This is why you had to work with the `culture` value before to get the English language descriptions. To make matters worse, you are told the sales are recorded in **local** currency, so your previous analysis is flawed, and you must convert all amounts to USD if you wish to compare the different salespeople fairly!

Use the `countryregioncurrency` table in combination with the `salesperson` and `salesterritory` ones to figure out the relevant currency symbol for each of the top salespeople.

**Answer.**

In [10]:
%%sql
WITH country_code AS (
    SELECT territoryid, countryregioncurrency.currencycode FROM salesterritory
    JOIN countryregioncurrency ON countryregioncurrency.countryregioncode = salesterritory.countryregioncode
), new_salesperson AS (
    SELECT businessentityid, country_code.currencycode FROM country_code
    JOIN salesperson ON salesperson.territoryid = country_code.territoryid
)

SELECT * FROM new_salesperson
LIMIT 20;

 * postgresql://postgres:***@ec3-database.cfnfrwra7yoo.us-east-2.rds.amazonaws.com/adventuretime
16 rows affected.


[(275, 'USD'),
 (276, 'USD'),
 (277, 'USD'),
 (278, 'CAD'),
 (279, 'USD'),
 (280, 'USD'),
 (281, 'USD'),
 (282, 'CAD'),
 (283, 'USD'),
 (284, 'USD'),
 (286, 'AUD'),
 (288, 'EUR'),
 (288, 'DEM'),
 (289, 'GBP'),
 (290, 'FRF'),
 (290, 'EUR')]

-------

### Exercise 11:

Now that we have the currency codes associated with each salesperson, redo Exercise 7 to take the currency exchange into account. If there are salespeople in the top 5 that weren't there before, explain why.

**Hint:** The rates in the```currencyrate``` table always go from `FromCurrencyCode=USD` to `ToCurrencyCode=<Desired Currency Code>`, and they are listed every day. When calculating line totals, use the `AverageRate` for that day. You should be able to reuse a lot of Exercise 7.

**Answer.**

The results from **exercise 10** show that there are two salespersons that made transactions in different currencies. For example, salesperson #288 had transactions denominated in deutsche marks and euros. This could be a potential problem for our current analysis since we could get duplicated orders. To illustrate this issue take for instance one of the orders made by this salesperson:

In [11]:
%%sql 
SELECT * FROM salesorderheader
WHERE salesorderid = '65089'

 * postgresql://postgres:***@ec3-database.cfnfrwra7yoo.us-east-2.rds.amazonaws.com/adventuretime
1 rows affected.


[(65089, 9, datetime.date(2014, 1, 28), datetime.date(2014, 2, 10), datetime.date(2014, 2, 5), 5, False, 'PO986112457', '10-4020-000518', 30079, 288, 8, 743, 743, 5, 206, '46635Vi1057', 11077, 1466.01, 125.658, 39.2681, 1630.9361, None, '91858e70-0b81-4e96-8cd1-a25a7a79ea3a', datetime.date(2014, 2, 5))]

As we can see, there is only one order that was carried out in Germany (`territoryid` == 8) and we know for a fact that its `currencyrateid` corresponds to EUR:  

In [12]:
%%sql
SELECT * FROM currencyrate
WHERE currencyrateid = '11077'

 * postgresql://postgres:***@ec3-database.cfnfrwra7yoo.us-east-2.rds.amazonaws.com/adventuretime
1 rows affected.


[(11077, datetime.date(2014, 1, 28), 'USD', 'EUR', 1.0085, 1.0084, datetime.date(2014, 1, 28))]

If we use the output of **exercise 10** to assign the currency code, we will duplicate the values of one single transaction:

In [13]:
%%sql
WITH country_code AS (
    SELECT territoryid, countryregioncurrency.currencycode FROM salesterritory
    JOIN countryregioncurrency ON countryregioncurrency.countryregioncode = salesterritory.countryregioncode
), new_salesperson AS (
    SELECT businessentityid, country_code.currencycode FROM country_code
    JOIN salesperson ON salesperson.territoryid = country_code.territoryid
), sales_2014 AS(
    SELECT salesorderid, orderdate,salespersonid,currencyrateid,subtotal,currencycode FROM salesorderheader
    JOIN new_salesperson ON new_salesperson.businessentityid = salesorderheader.salespersonid
    WHERE (TO_CHAR(orderdate,'YYYY') = '2014' AND salespersonid IS NOT NULL)
)

SELECT * FROM sales_2014
WHERE salesorderid = '65089'

 * postgresql://postgres:***@ec3-database.cfnfrwra7yoo.us-east-2.rds.amazonaws.com/adventuretime
2 rows affected.


[(65089, datetime.date(2014, 1, 28), 288, 11077, 1466.01, 'DEM'),
 (65089, datetime.date(2014, 1, 28), 288, 11077, 1466.01, 'EUR')]

Now, given that we have `currencyrateid` field on the `salesorderheader` table, why do we bother extracting the currency code from other tables? Because there are a lot of transactions that doesn't have an appropiate identification (for 2014 alone there are more than 6.000 null observations of this kind):  

In [14]:
%%sql 
SELECT count(*) FROM salesorderheader
WHERE (currencyrateid IS NULL)
LIMIT 10

 * postgresql://postgres:***@ec3-database.cfnfrwra7yoo.us-east-2.rds.amazonaws.com/adventuretime
1 rows affected.


[(17489,)]

Problem that could easily be solved when assigning a currency code to each salesperson:

In [15]:
%%sql
WITH country_code AS (
    SELECT territoryid, countryregioncurrency.currencycode FROM salesterritory
    JOIN countryregioncurrency ON countryregioncurrency.countryregioncode = salesterritory.countryregioncode
), new_salesperson AS (
    SELECT businessentityid, country_code.currencycode FROM country_code
    JOIN salesperson ON salesperson.territoryid = country_code.territoryid
), sales_2014 AS(
    SELECT salesorderid, orderdate,salespersonid,currencyrateid,subtotal,currencycode FROM salesorderheader
    JOIN new_salesperson ON new_salesperson.businessentityid = salesorderheader.salespersonid
    WHERE (TO_CHAR(orderdate,'YYYY') = '2014' AND salespersonid IS NOT NULL)
)

SELECT count(*) FROM sales_2014
WHERE (currencycode IS NULL)

 * postgresql://postgres:***@ec3-database.cfnfrwra7yoo.us-east-2.rds.amazonaws.com/adventuretime
1 rows affected.


[(0,)]

Since the problem is asking us to replicate the **exercise 7** taking into account the currency code, we know that we could do the exercise on a subset of data. Specifically, we'd like to determine the top-performing salesperson **during 2014**. This is a valuable piece of information, since we know that since 2011 the EUR is the currency adopted by many countries in Europe, included Germany and France. For this reason, we could modify the snippet of code posted before to select the observations that includes USD, EUR, GBP, CAD and AUD only.

In [16]:
%%sql
WITH country_code AS (
    SELECT territoryid, countryregioncurrency.currencycode FROM salesterritory
    JOIN countryregioncurrency ON countryregioncurrency.countryregioncode = salesterritory.countryregioncode
), new_salesperson AS (
    SELECT businessentityid, country_code.currencycode FROM country_code
    JOIN salesperson ON salesperson.territoryid = country_code.territoryid
), sales_2014 AS(
    SELECT salesorderid, orderdate,salespersonid,currencyrateid,subtotal,currencycode FROM salesorderheader
    JOIN new_salesperson ON new_salesperson.businessentityid = salesorderheader.salespersonid
    WHERE (TO_CHAR(orderdate,'YYYY') = '2014' AND salespersonid IS NOT NULL AND currencycode!='DEM' AND currencycode!='FRF')
), sales_to_currency AS (
    SELECT orderdate,sales_2014.salespersonid,subtotal,currencycode,averagerate,(COALESCE((unitprice * (1-unitpricediscount) * orderqty),0)/averagerate) AS linetotal_USD FROM sales_2014
    JOIN salesorderdetail ON salesorderdetail.salesorderid = sales_2014.salesorderid
    JOIN currencyrate ON (currencyrate.currencyratedate = sales_2014.orderdate AND currencyrate.tocurrencycode = sales_2014.currencycode)
)

SELECT salespersonid, SUM(linetotal_USD) AS tot_sales_USD FROM sales_to_currency
GROUP BY salespersonid
ORDER BY tot_sales_USD DESC
LIMIT 5;

 * postgresql://postgres:***@ec3-database.cfnfrwra7yoo.us-east-2.rds.amazonaws.com/adventuretime
5 rows affected.


[(289, 2146418.4023947986),
 (276, 1271088.5214609983),
 (275, 1057247.3785719983),
 (277, 1040093.4069009982),
 (290, 844392.7295821798)]

Comparing the results to those of **exercise 7** we could see that the fourth and fifth salespersons change in this exercise, where #277 scale one position and #290 enters the list. The reason behind this is the adjustment of sales by currency as explained before. For example, #290 sold items denominated in EUR, while #282 (who was eight with the adjustment) sold items denominated in CAD.

-------

### Exercise 12:

How does the correlation from Exercise 9 change once you've adjusted for the currency?

**Answer.**

In [17]:
%%sql
WITH country_code AS (
    SELECT territoryid, countryregioncurrency.currencycode FROM salesterritory
    JOIN countryregioncurrency ON countryregioncurrency.countryregioncode = salesterritory.countryregioncode
), new_salesperson AS (
    SELECT businessentityid, country_code.currencycode FROM country_code
    JOIN salesperson ON salesperson.territoryid = country_code.territoryid
), sales_2014 AS(
    SELECT salesorderid, orderdate,salespersonid,currencyrateid,subtotal,currencycode FROM salesorderheader
    JOIN new_salesperson ON new_salesperson.businessentityid = salesorderheader.salespersonid
    WHERE (TO_CHAR(orderdate,'YYYY') = '2014' AND salespersonid IS NOT NULL AND currencycode!='DEM' AND currencycode!='FRF')
), sales_to_currency AS (
    SELECT orderdate,sales_2014.salespersonid,subtotal,currencycode,averagerate,(COALESCE((unitprice * (1-unitpricediscount) * orderqty),0)/averagerate) AS linetotal_USD FROM sales_2014
    JOIN salesorderdetail ON salesorderdetail.salesorderid = sales_2014.salesorderid
    JOIN currencyrate ON (currencyrate.currencyratedate = sales_2014.orderdate AND currencyrate.tocurrencycode = sales_2014.currencycode)
)


SELECT salespersonid, corr(linetotal_usd,commissionpct) FROM sales_to_currency
JOIN salesperson ON salesperson.businessentityid = sales_to_currency.salespersonid
GROUP BY salespersonid
ORDER BY corr DESC
LIMIT 20;

 * postgresql://postgres:***@ec3-database.cfnfrwra7yoo.us-east-2.rds.amazonaws.com/adventuretime
14 rows affected.


[(286, 0.12609611927979364),
 (279, 0.08016705326788381),
 (290, 0.034673529991778966),
 (276, 0.007533553197528419),
 (275, 0.005194760002641503),
 (277, 0.0032238322840994235),
 (289, -0.003515573146075402),
 (283, -0.014686665679863799),
 (288, -0.02007638628090017),
 (278, -0.028112294774936316),
 (281, -0.03396659488291904),
 (280, -0.05407034018559087),
 (284, -0.05805456773100009),
 (282, -0.10115459510615189)]

As we can see, there are some major changes on the correlation values from **exercise 9** to this one. For instance, the salesperson #286, which had a correlation of 1% before adjusting by currency, exhibits the greatest value among all salespersons (12,6%). In other cases the correlation stays negative, but is lower when we adjust for currency rate (e.g. #281 passes from -9,6% to -3,3%). 

-------

## Attribution

"AdventureWorks database", Nov 7, 2017, Microsoft Corporation, [MIT License](https://docs.microsoft.com/en-us/sql/samples/sql-samples-where-are?view=sql-server-ver15), https://github.com/microsoft/sql-server-samples/tree/master/samples/databases/adventure-works